datafusion/sql/src/planner.rs

```diff
@@ -970,6 +970,46 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         }
     }
 
+    /// Find all `PlaceHolder` tokens in a logical plan, and try to infer their type from context
+    fn infer_placeholder_types(&self, expr: &mut Expr, schema: &DFSchema) -> Result<()> {
+        match expr {
+            Expr::BinaryExpr(BinaryExpr {
+                ref mut left,
+                ref mut right,
+                ..
+            }) => {
+                self.infer_placeholder_types(left, schema)?;
+                self.infer_placeholder_types(right, schema)?;
+                let lt = left.get_type(schema);
+                let rt = right.get_type(schema);
+                match (left, rt) {
```

Comment:
I was thinking in general this logic is only correct for "comparison" binary operators. It works well for

```
a > $1
```

However, I couldn't come up with an example where it wouldn't work (even `Operator::Like` would result in the other argument being a string.